sdks/java/extensions/kryo/src/main/java/org/apache/beam/sdk/extensions/kryo/KryoCoderProvider.java

```diff
@@ -0,0 +1,144 @@
+
+import static java.util.Objects.requireNonNull;
+
+import com.esotericsoftware.kryo.ClassResolver;
+import com.esotericsoftware.kryo.Kryo;
+import com.esotericsoftware.kryo.Registration;
+import com.google.common.annotations.VisibleForTesting;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.apache.beam.sdk.Pipeline;
+import org.apache.beam.sdk.coders.CannotProvideCoderException;
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.coders.CoderProvider;
+import org.apache.beam.sdk.values.TypeDescriptor;
+
+/**
+ * Implementation of {@link CoderProvider}, which provides {@link KryoCoder} for any type registered
+ * to {@link Kryo} by user-provided {@link KryoRegistrar}.
+ */
+public class KryoCoderProvider extends CoderProvider {
+
+  private final Map<TypeDescriptor, KryoCoder<?>> typeToCoder = new HashMap<>();
+  private final IdentifiedRegistrar kryoRegistrar;
+
+  /**
+   * Starts build of {@link KryoCoderProvider} with given {@link KryoRegistrar}.
+   *
+   * @param registrar user defined implementation of {@link KryoRegistrar}
+   * @return next step of building
+   */
+  public static FinalBuilder of(KryoRegistrar registrar) {
+    return new Builder(registrar);
+  }
+
+  private KryoCoderProvider(IdentifiedRegistrar kryoRegistrar) {
+    requireNonNull(kryoRegistrar);
+    this.kryoRegistrar = kryoRegistrar;
+  }
+
+  @Override
+  public <T> Coder<T> coderFor(
+      TypeDescriptor<T> typeDescriptor, List<? extends Coder<?>> componentCoders)
+      throws CannotProvideCoderException {
+
+    // try to obtain cached KryoCoder coder by type descriptor
+    KryoCoder<?> coder = typeToCoder.get(typeDescriptor);
+
+    // if not cached, try to obtain coder for type registered in kryo
+    if (coder == null) {
+      coder = createKryoCoderIfRawTypeRegistered(typeDescriptor);
+    }
+
+    @SuppressWarnings("unchecked")
+    Coder<T> castedCoder = (Coder<T>) coder;
+
+    return castedCoder;
+  }
+
+  private <T> KryoCoder<T> createKryoCoderIfRawTypeRegistered(TypeDescriptor<T> typeDescriptor)
+      throws CannotProvideCoderException {
+
+    Class<? super T> rawType = typeDescriptor.getRawType();
+    Kryo kryo = KryoFactory.getOrCreateKryo(kryoRegistrar);
+    ClassResolver classResolver = kryo.getClassResolver();
+
+    Registration registration = classResolver.getRegistration(rawType);
+    if (registration == null) {
+      throw new CannotProvideCoderException(
+          String.format(
+              "Cannot provide %s, given type descriptor's '%s' raw type is not registered in Kryo.",
+              KryoCoder.class.getSimpleName(), typeDescriptor));
+    }
+
+    return KryoCoder.of(kryoRegistrar);
+  }
+
+  @VisibleForTesting
+  IdentifiedRegistrar getKryoRegistrar() {
+    return kryoRegistrar;
+  }
+
+  // ----------------- Builder steps
+
+  /** Last step when building {@link KryoCoderProvider}. */
+  public interface FinalBuilder {
```

Comment:
Do we need the builder at all? It seems that it does not have any added value in this case, simple `KryoCoderProvider.of(registrar)` should be sufficient.

Reply:
I believe that mot of the time (apart from tests) user will want to create and register `KryoProvider`. So we can either provide `of(registrar, pipeline)` or `of(registrar).buildAndRegister(pipeline)`. I favor the second option, since `of(registrar, pipeline)` would require user to read docs to understand what it does.